### PR TITLE
Remove DisallowNullAttribute from base equality comparers

### LINKED ISSUE #44: 'AllowNullAttribute::.ctor()' is declared in another module and needs to be imported
@sharwell We missed something in https://github.com/tunnelvisionlabs/ReferenceAssemblyAnnotator/pull/43.

```
The "AnnotatorBuildTask" task failed unexpectedly.
System.ArgumentException: Member 'System.Void System.Diagnostics.CodeAnalysis.AllowNullAttribute::.ctor()' is declared in another module and needs to be imported
   at Mono.Cecil.MetadataBuilder.LookupToken(IMetadataTokenProvider provider)
   at Mono.Cecil.MetadataBuilder.AddCustomAttributes(ICustomAttributeProvider owner)
   at Mono.Cecil.MetadataBuilder.AddParameter(UInt16 sequence, ParameterDefinition parameter, ParamTable table)
   at Mono.Cecil.MetadataBuilder.AddParameters(MethodDefinition method)
   at Mono.Cecil.MetadataBuilder.AddMethod(MethodDefinition method)
   at Mono.Cecil.MetadataBuilder.AddMethods(TypeDefinition type)
   at Mono.Cecil.MetadataBuilder.AddType(TypeDefinition type)
   at Mono.Cecil.MetadataBuilder.AddTypes()
   at Mono.Cecil.MetadataBuilder.BuildTypes()
   at Mono.Cecil.MetadataBuilder.BuildModule()
   at Mono.Cecil.MetadataBuilder.BuildMetadata()
   at Mono.Cecil.ModuleWriter.<>c.<BuildMetadata>b__2_0(MetadataBuilder builder, MetadataReader _)
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TItem item, Func`3 read)
   at Mono.Cecil.ModuleWriter.BuildMetadata(ModuleDefinition module, MetadataBuilder metadata)
   at Mono.Cecil.ModuleWriter.Write(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters)
   at Mono.Cecil.ModuleWriter.WriteModule(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters)
   at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters)
   at TunnelVisionLabs.ReferenceAssemblyAnnotator.Program.Main(Nullable`1 log, String referenceAssembly, String annotatedReferenceAssembly, String outputAssembly)
   at TunnelVisionLabs.ReferenceAssemblyAnnotator.AnnotatorBuildTask.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
```

https://dev.azure.com/guorg/Gu.Roslyn.Extensions/_build/results?buildId=332&view=logs&jobId=12f1170f-54f2-53f3-20dd-22fc7dff55f9&taskId=006e6f7c-8314-5e02-a3df-483f3a1f4cfa&lineStart=15&lineEnd=37&colStart=1&colEnd=474

### LINKED ISSUE #46: Remove [DisallowNull] on IEqualityComparer<T>.GetHashCode
These `[DisallowNull]` attributes should be removed:

* `IEqualityComparer<T>.GetHashCode(T)`
* `EqualityComparer<T>.GetHashCode(T)`

The following scenarios are impacted:

* For netcoreapp3.0 and newer targets, the reference assemblies should be rewritten to remove the existing attribute from the reference assembly
* For earlier targets, the attributes should simply not be added during the rewrite process

### PR DESCRIPTION
**Pull request by sharwell**
Works around dotnet/corefx#41405 by substituting a new reality

Fixes #44
Closes #46

### CODE REVIEW

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs [hunk 1]
```diff
@@ -207,6 +220,25 @@ static bool IsMatchingConstructor(MethodDefinition constructor, CustomAttribute
...
+            if (provider is ParameterDefinition parameter
+                && parameter.Method is MethodDefinition { Name: nameof(GetHashCode), Parameters: { Count: 1 } } method
+                && method.DeclaringType is { Namespace: "System.Collections.Generic", GenericParameters: { Count: 1 } } type)
+            {
+                if (type.Name == "IEqualityComparer`1"
+                    || type.Name == "EqualityComparer`1")
+                {
+                    // Remove DisallowNullAttribute from:
+                    // 1. System.Collections.Generic.IEqualityComparer<T>.GetHashCode([DisallowNull] T)
+                    // 2. System.Collections.Generic.EqualityComparer<T>.GetHashCode([DisallowNull] T)
```
Comment (jnm2) [inline TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs:234]: Why not just remove from #\2?

Comment (jnm2) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs:234]: (I typed `\#2`. It hid the backslash without escaping anything. Didn't that once escape # links?)

Comment (sharwell) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs:234]: The interface should not be preventing a user from implementing an equality comparer for a value domain where `null` is a valid value.

Comment (jnm2) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs:234]: But it wouldn't be preventing the user from implementing such an equality comparer. The user can always widen it for their class, just like you'd be doing to the EqualityComparer class by removing its DisallowNull even if you didn't remove the DisallowNull from IEqualityComparer.

Comment (sharwell) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs:234]: Except it does prevent it. A type should be able to have `IEqualityComparer<MyObject?>` as an input, knowing that the type needs a comparer capable of operating on any `MyObject` instance or `null`, and it should not be required to be tightly coupled to one implementation of the interface in order to do so.

Comment (jnm2) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs:234]: Ahh. I see how it prevents consumption of `IEqualityComparer` while allowing null, not how it prevents implementation (per se) of `IEqualityComparer` while allowing null, but I think I'm following.

**Pull request merged**